panic while unmarshall (go-yaml/yaml issue #1050)

https://github.com/go-yaml/yaml/issues/1050 applies to the fork:

```Go
package main

import (
	"fmt"

	yaml "sigs.k8s.io/yaml/goyaml.v3"
)

type SomeStruct struct {
	data string `yaml:"data"`
}

func main() {
	someStruct := SomeStruct{}

	var data = []byte{
		0x34, 0x30, 0x79, 0x3a,
		0x3a, 0x0d, 0x0d, 0x0d,
		0x0d, 0x0d, 0x0d, 0x0d,
		0x0d, 0x3c, 0x3c, 0x3a,
		0x0d, 0x0d, 0x0d, 0x0d,
		0x0d, 0x0d, 0x0d, 0x2d,
		0x20, 0x20, 0x2d, 0x20,
		0x23, 0x0d, 0x3f, 0x0d,
		0x23, 0x0d, 0x0d, 0x23,
		0x2d, 0x0d, 0x20, 0x2d,
		0x0d, 0x3f, 0x0d, 0x23,
		0x0d, 0x0d, 0x23, 0x2d,
		0x2d,
	}
	fmt.Printf("%q\n", string(data))

	yaml.Unmarshal(data, &someStruct)
}
```

```
"40y::\r\r\r\r\r\r\r\r<<:\r\r\r\r\r\r\r-  - #\r?\r#\r\r#-\r -\r?\r#\r\r#--"
panic: runtime error: hash of unhashable type []interface {} [recovered]
	panic: runtime error: hash of unhashable type []interface {}

goroutine 1 [running]:
sigs.k8s.io/yaml/goyaml%2ev3.handleErr(0xc00014de78)
	/nvme/gopath/src/k8s.io/kubernetes/vendor/sigs.k8s.io/yaml/goyaml.v3/yaml.go:294 +0x6d
panic({0x507f80?, 0xc00011c230?})
	/nvme/gopath/go-1.23.0/src/runtime/panic.go:785 +0x132
sigs.k8s.io/yaml/goyaml%2ev3.(*decoder).merge(0xc0001435e0, 0xc000150320, 0xc0001505a0, {0x509d20?, 0xc00011c120?, 0x4e76e5?})
	/nvme/gopath/src/k8s.io/kubernetes/vendor/sigs.k8s.io/yaml/goyaml.v3/decode.go:966 +0x1a5
sigs.k8s.io/yaml/goyaml%2ev3.(*decoder).mappingStruct(0xc0001435e0, 0xc000150320, {0x509d20?, 0xc00011c120?, 0xc00014db10?})
	/nvme/gopath/src/k8s.io/kubernetes/vendor/sigs.k8s.io/yaml/goyaml.v3/decode.go:950 +0x1065
sigs.k8s.io/yaml/goyaml%2ev3.(*decoder).mapping(0xc0001435e0, 0xc000150320, {0x509d20?, 0xc00011c120?, 0x46251e?})
	/nvme/gopath/src/k8s.io/kubernetes/vendor/sigs.k8s.io/yaml/goyaml.v3/decode.go:786 +0xa5
sigs.k8s.io/yaml/goyaml%2ev3.(*decoder).unmarshal(0xc0001435e0, 0xc000150320, {0x509d20?, 0xc00011c120?, 0x4d4385?})
	/nvme/gopath/src/k8s.io/kubernetes/vendor/sigs.k8s.io/yaml/goyaml.v3/decode.go:510 +0x3ea
sigs.k8s.io/yaml/goyaml%2ev3.(*decoder).document(...)
	/nvme/gopath/src/k8s.io/kubernetes/vendor/sigs.k8s.io/yaml/goyaml.v3/decode.go:527
sigs.k8s.io/yaml/goyaml%2ev3.(*decoder).unmarshal(0xc0001435e0, 0xc000150280, {0x509d20?, 0xc00011c120?, 0xc00014de78?})
	/nvme/gopath/src/k8s.io/kubernetes/vendor/sigs.k8s.io/yaml/goyaml.v3/decode.go:498 +0x28a
sigs.k8s.io/yaml/goyaml%2ev3.unmarshal({0xc00016c000, 0x31, 0x31}, {0x4fd8c0, 0xc00011c120}, 0x1?)
	/nvme/gopath/src/k8s.io/kubernetes/vendor/sigs.k8s.io/yaml/goyaml.v3/yaml.go:167 +0x396
sigs.k8s.io/yaml/goyaml%2ev3.Unmarshal(...)
	/nvme/gopath/src/k8s.io/kubernetes/vendor/sigs.k8s.io/yaml/goyaml.v3/yaml.go:89
main.main()
	/nvme/gopath/src/k8s.io/kubernetes/test2/main.go:33 +0x107
exit status 2
```

--- a/goyaml.v2/patch_test.go
+++ b/goyaml.v2/patch_test.go
@@ -1,0 +1,23 @@
+package yaml_test
+
+import (
+	"testing"
+
+	yaml "sigs.k8s.io/yaml/goyaml.v2"
+)
+
+func TestIssue117(t *testing.T) {
+	data := []byte(`
+a:
+<<:
+-
+?
+-
+`)
+
+	x := map[string]interface{}{}
+	err := yaml.Unmarshal([]byte(data), &x)
+	if err == nil {
+		t.Errorf("expected error, got none")
+	}
+}

--- a/goyaml.v3/decode.go
+++ b/goyaml.v3/decode.go
@@ -832,10 +832,10 @@ func (d *decoder) mapping(n *Node, out reflect.Value) (good bool) {
 		if d.unmarshal(n.Content[i], k) {
 			if mergedFields != nil {
 				ki := k.Interface()
-				if mergedFields[ki] {
+				if d.getPossiblyUnhashableKey(mergedFields, ki) {
 					continue
 				}
-				mergedFields[ki] = true
+				d.setPossiblyUnhashableKey(mergedFields, ki, true)
 			}
 			kkind := k.Kind()
 			if kkind == reflect.Interface {
@@ -956,6 +956,24 @@ func failWantMap() {
 	failf("map merge requires map or sequence of maps as the value")
 }
 
+func (d *decoder) setPossiblyUnhashableKey(m map[interface{}]bool, key interface{}, value bool) {
+	defer func() {
+		if err := recover(); err != nil {
+			failf("%v", err)
+		}
+	}()
+	m[key] = value
+}
+
+func (d *decoder) getPossiblyUnhashableKey(m map[interface{}]bool, key interface{}) bool {
+	defer func() {
+		if err := recover(); err != nil {
+			failf("%v", err)
+		}
+	}()
+	return m[key]
+}
+
 func (d *decoder) merge(parent *Node, merge *Node, out reflect.Value) {
 	mergedFields := d.mergedFields
 	if mergedFields == nil {
@@ -963,7 +981,7 @@ func (d *decoder) merge(parent *Node, merge *Node, out reflect.Value) {
 		for i := 0; i < len(parent.Content); i += 2 {
 			k := reflect.New(ifaceType).Elem()
 			if d.unmarshal(parent.Content[i], k) {
-				d.mergedFields[k.Interface()] = true
+				d.setPossiblyUnhashableKey(d.mergedFields, k.Interface(), true)
 			}
 		}
 	}

--- a/goyaml.v3/patch_test.go
+++ b/goyaml.v3/patch_test.go
@@ -18,6 +18,7 @@ package yaml_test
 
 import (
 	"bytes"
+	"testing"
 
 	. "gopkg.in/check.v1"
 	yaml "sigs.k8s.io/yaml/goyaml.v3"
@@ -157,4 +158,20 @@ func (s *S) TestNewLinePreserved(c *C) {
 	c.Assert(err, IsNil)
 	// the newline at the start of the file should be preserved
 	c.Assert(string(data), Equals, "_: |4\n\n    a:\n            b:\n                    c: d\n")
+}
+
+func TestIssue117(t *testing.T) {
+	data := []byte(`
+a:
+<<:
+-
+?
+-
+`)
+
+	x := map[string]interface{}{}
+	err := yaml.Unmarshal([]byte(data), &x)
+	if err == nil {
+		t.Errorf("expected error, got none")
+	}
 }

